File: Platforms/Common/Reign.Video.XNA/States/BlendState.cs
using Reign.Core;
using X = Microsoft.Xna.Framework.Graphics;

namespace Reign.Video.XNA
{
	public class BlendState : Disposable, BlendStateI
	{
		#region Properties
		private Video video;
		private BlendStateDesc desc;
		#endregion

		#region Constructors
		public static BlendState New(DisposableI parent, BlendStateDescI desc)
		{
			return new BlendState(parent, desc);
		}

		public BlendState(DisposableI parent, BlendStateDescI desc)
		: base(parent)
		{
			video = parent.FindParentOrSelfWithException<Video>();
			this.desc = (BlendStateDesc)desc;
		}
		#endregion

		#region Methods
		public void Enable()
		{
			if (desc.blendEnable)
			{
				var blendState = new X.BlendState()
				{
					ColorWriteChannels = desc.renderTargetWriteMask,
					
					ColorBlendFunction = desc.blendOp,
					ColorSourceBlend = desc.srcBlend,
					ColorDestinationBlend = desc.dstBlend,

					AlphaBlendFunction = desc.blendOpAlpha,
					AlphaSourceBlend = desc.srcBlendAlpha,
					AlphaDestinationBlend = desc.dstBlendAlpha
				};
			
				video.Device.BlendState = blendState;
			}
			else
			{
				video.Device.BlendState = X.BlendState.Opaque;
			}
		}
		#endregion
	}
}

File: Platforms/Windows/Reign_Video_D3D9_Component/BlendState.cpp
#include "pch.h"
#include "BlendState.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	BlendStateCom::BlendStateCom(VideoCom^ video, BlendStateDescCom^ desc)
	{
		this->video = video;
		this->desc = desc;
	}
	#pragma endregion
	
	#pragma region Methods
	void BlendStateCom::Enable()
	{
		video->device->SetRenderState(D3DRS_COLORWRITEENABLE, desc->renderTargetWriteMask);

		video->device->SetRenderState(D3DRS_ALPHABLENDENABLE, desc->blendEnable);
		if (desc->blendEnable)
		{
			video->device->SetRenderState(D3DRS_BLENDOP, desc->blendOp);
			video->device->SetRenderState(D3DRS_SRCBLEND, desc->srcBlend);
			video->device->SetRenderState(D3DRS_DESTBLEND, desc->dstBlend);
		}

		video->device->SetRenderState(D3DRS_SEPARATEALPHABLENDENABLE, desc->blendEnableAlpha);
		if (desc->blendEnableAlpha)
		{
			video->device->SetRenderState(D3DRS_BLENDOPALPHA, desc->blendOpAlpha);
			video->device->SetRenderState(D3DRS_SRCBLENDALPHA, desc->srcBlendAlpha);
			video->device->SetRenderState(D3DRS_DESTBLENDALPHA, desc->dstBlendAlpha);
		}
	}
	#pragma endregion
}